doc/specs/#140 - ZIP Support.md

```diff
@@ -0,0 +1,135 @@
+---
+author: Alexis Bekhdadi @midoriiro
+created on: 2021-11-16
+last updated: 2021-11-16
+issue id: 140
+---
+
+# Spec Title
+
+For [#140](https://github.com/microsoft/winget-cli/issues/140)
+
+## Abstract
+
+This feature adds to winget-cli the ability to install application delivered through a ZIP archive.
+
+## Inspiration
+
+A bunch of applications are only provided by developer's through a ZIP archive. There are three cases:
+1) The archive contain only one binary that need do to be extracted and add to PATH env var
+2) The archive contain an installer that need to be executed and let do his job
+3) The archive contain a complex hierarchy that need do to be extracted and add specific folder to PATH env var
+
+## Solution Design
+
+A new installation flow need to be created to execute different installation strategies regarding these three cases above.
+
+First thing first, the archive need to be extracted in a specific path (TEMP folder can be a good candidate).
+
+Regarding case 1, installation flow check if TEMP folder contain a single binary and repack it
+into an MSI package (including setting PATH env var). Then, MSI installation flow can take control.
```

Comment:
Packaging a standalone / portable application is an interesting approach. I'm not sure we had considered this specific approach. We may prefer the concept of repackaging as an MSIX as that is the latest technology, but there may be different concerns to discuss here.

Reply:
I'm not sure that writing code to transform the archive into an installable type is better than simply writing the code to manage archives directly.  All of the proposals require unpacking the archive, which is itself the primary step in management.  Afterward, adding to the `%PATH%` is fairly straightforward and the issues are more about lifetime management.  Injecting our own ARP entry is largely sufficient to achieve that.

Reply:
@denelon I'm not an expert concerning Windows packaging application. The main idea of this specification is to package the content of a zip file (except for case 2) in order to let winget managing the lifecycle of the new package. 

I saw Windows SDK offers a set of CLI command to create MSIX package, but that mean installing Windows SDK.

@JohnMcPMS  I don't know what is ARP, but if I understood your suggestion. How you do you handle uninstall/upgrade of the package ? That the goal of an ARP entry ?

Reply:
Add Remove Programs (ARP) is where many of the entries for programs installed on Windows 10 and Windows 11 get recorded. We use that metadata for the list command and as the source for values we used to correlate with manifests in configured sources. We don't necessarily need to create an "insaller" per se'. It's possible to just write entries to ARP (now part of Apps & Features). The Windows Package Manager could act as an "installer" proxy to create the entries and modify the path environment variable. We're looking at pros and cons to several approaches.

Reply:
I see, it's make sense. I looked if an API exist about managing an ARP entry, but I did found nothing. If winget-cli already manage that kind of tasks, @JohnMcPMS made a good point.

I don't know what is the state of art about ARP in Windows 10/11 now. But, ten years ago I wrote an application that find all installed applications on the system through the registry, the main goal was to uninstall silently a batch of applications. That was tricky to handle all the cases. 

If an API already exist, the repacking solution would not be the best approach(in my opinion) as you explained both of you.

I'll try to find some documentation on the internet and this repository to study how to manage an ARP entry.

Reply:
I don't think that there are APIs; we certainly do not use them and I looked for them when we implemented reading.  It is basically a registry data contract.

Reply:
Yes, I saw that yesterday. I updated the spec. There some points to discuss about it but, especially where to install the zip content. The ARP section seem good for you ?

Reply:
Perhaps we treat ZIP installs the same way visual studio code is installed (i.e. user installer) within the AppData folder.

Reply:
That I suggested in the last revision of the spec. For user scope, application will be installed in AppData/Local/myapplication. But maybe there is a better location for user scope ? 

In any ways, scope and location can be overrides by users when calling winget install. Just found that, that should be take in consideration on the specification.